### PR TITLE
roachpb: handle MVCC extended value encoding in roachpb.Value

### PR DESCRIPTION
**Pull request by stevendanna**
Previously, values encoded with the extended encoding would never
escape pkg/storage.

Now, requests made with the new ReturnRawMVCCValues option will return
roachpb.Value structs with their RawBytes containing the full extended
encoded value.

Here, we extended the following method on roachpb.Value to handle this
encoding:

- GetTag: Reads past the value header and returns the underlying tag.

- Get[Type]: Reads past the value header and returns the underlying
  data.

- IsPresent(): Now additionally returns false if RawBytes is non-nil but _only_
  contains the MVCC Value header.

- TagAndDataBytes(): Reads past the value header and returns the
  underlying tag and data.

- checksum/computeCheckum/Verify(): Reads past the value header to return/compute the checksum.

We additionally add GetMVCCValueHeader which returns the
MVCCValueHeader if one exists.

Epic: none
Release note: None

### CODE REVIEW

File: pkg/roachpb/data.go [hunk 1]
```diff
@@ -346,7 +363,20 @@ func (v *Value) ShallowClone() *Value {
...
-	return v != nil && len(v.RawBytes) != 0
+	if v == nil || len(v.RawBytes) == 0 {
+		return false
+	}
+	// TODO(ssd): This is a bit awkward because this is the right thing to
+	// do for production callers trying to determine if this value is a
+	// tombstone. But, many tests shove random strings into RawBytes, and in
+	// then case we'll hit this case if the 5th character of that string
+	// happens to be `e` (ascii 101). There aren't _that_ many callers to
+	// IsPresent(). We may just need to audit them all.
```
Comment (stevendanna) [inline pkg/roachpb/data.go:374]: On the larger PR, @tbg says:

> Can you address this TODO before merging?

**Review by miraradeva — APPROVED** [review]
:lgtm: Just a few nits and questions.

> Reviewed 4 of 4 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132088#-:-O8bQZdQAofJdY_-AwxI:b-jvaqn0)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @miraradeva and @stevendanna)

___
*[`-- commits` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132088#-O8bIPbtEbLkhcf24hYb):*
nit: "methods"
___
*[`-- commits` line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132088#-O8bIRSj9S6YhclyZhYB):*
Would this case (empty `RawBytes` and non-empty `MVCCValueHeader`) be a tombstone with a `MVCCValueHeader`?
___
*[`pkg/roachpb/data.go` line 367 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132088#-O8bLJaEAEc0QUYqxFUH:-O8bLJaEAEc0QUYqxFUI:bdkpfgp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294e8d61941704997aa8bc5049b4d8de3bc1d2b/pkg/roachpb/data.go#L367)):*
> ```Go
> 	// IsPresent(). We may just need to audit them all.
> 	if len(v.RawBytes) >= extendedPreludeSize && v.RawBytes[tagPos] == byte(ValueType_MVCC_EXTENDED_ENCODING_SENTINEL) {
> 		headerSize := extendedPreludeSize + binary.BigEndian.Uint32(v.RawBytes)
> ```

nit: maybe use a different variable name so it's not confused with the `headerSize` defined above?
___
*[`pkg/roachpb/data.go` line 370 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132088#-O8bKe1CA9pLwcrbwa5k:-O8bKe1CA9pLwcrbwa5l:bffcg0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294e8d61941704997aa8bc5049b4d8de3bc1d2b/pkg/roachpb/data.go#L370)):*
> ```Go
> 		return len(v.RawBytes) > int(headerSize)
> 	}
> 	return len(v.RawBytes) >= 0
> ```

We know `len(v.RawBytes) > 0`, so return `true`?
___
*[`pkg/roachpb/data.go` line 417 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132088#-O8bMMIG88KF72eePpX_:-O8bMMIG88KF72eePpXa:b-htvzzc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294e8d61941704997aa8bc5049b4d8de3bc1d2b/pkg/roachpb/data.go#L417)):*
> ```Go
> 	if v.RawBytes[tagPos] == byte(ValueType_MVCC_EXTENDED_ENCODING_SENTINEL) {
> 		headerLen := binary.BigEndian.Uint32(v.RawBytes)
> 		headerSize := extendedPreludeSize + headerLen
> ```

Same nit about the variable name. Maybe call it `extendedHeaderSize`?
___
*[`pkg/gossip/util_test.go` line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132088#-O8bNBXO6CUQg32o4BBD:-O8bNBXO6CUQg32o4BBE:b-rjyll6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294e8d61941704997aa8bc5049b4d8de3bc1d2b/pkg/gossip/util_test.go#L46)):*
> ```Go
> 	}
> 	rawBytes := randutil.RandBytes(rng, 100)
> 	// Ensure we don't accidentally make this look like an extended encoding sentinal.
> ```

nit: "sentinel"

**Review by stevendanna — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132088)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @miraradeva)

___
*[`-- commits` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132088#-O8bIPbtEbLkhcf24hYb):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

nit: "methods"
</blockquote></details>

Thanks, I'll fix this right before merging. Trying to avoid rebasing too often because I have a few branches in flight.
___
*[`-- commits` line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132088#-O8bIRSj9S6YhclyZhYB):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Would this case (empty `RawBytes` and non-empty `MVCCValueHeader`) be a tombstone with a `MVCCValueHeader`?
</blockquote></details>

Yup, exactly.
___
*[`pkg/roachpb/data.go` line 367 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132088#-O8bLJaEAEc0QUYqxFUH:-O8c1Lyq63mjADFXYiZ8:bcvjgzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294e8d61941704997aa8bc5049b4d8de3bc1d2b/pkg/roachpb/data.go#L367)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

nit: maybe use a different variable name so it's not confused with the `headerSize` defined above?
</blockquote></details>

Good idea. Done.
___
*[`pkg/roachpb/data.go` line 370 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132088#-O8bKe1CA9pLwcrbwa5k:-O8c1NmmDnF3xHzgNKdg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294e8d61941704997aa8bc5049b4d8de3bc1d2b/pkg/roachpb/data.go#L370)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

We know `len(v.RawBytes) > 0`, so return `true`?
</blockquote></details>

Done.
___
*[`pkg/roachpb/data.go` line 417 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132088#-O8bMMIG88KF72eePpX_:-O8c1QIO0nSlbR9g0Y_D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294e8d61941704997aa8bc5049b4d8de3bc1d2b/pkg/roachpb/data.go#L417)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Same nit about the variable name. Maybe call it `extendedHeaderSize`?
</blockquote></details>

Done.
___
*[`pkg/gossip/util_test.go` line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132088#-O8bNBXO6CUQg32o4BBD:-O8c1RvLE_8Lge47HbUh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294e8d61941704997aa8bc5049b4d8de3bc1d2b/pkg/gossip/util_test.go#L46)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

nit: "sentinel"
</blockquote></details>

Done.

File: pkg/roachpb/data.go (hunk 1 above)
Comment (stevendanna) [reply pkg/roachpb/data.go:374]: I can see how far we can get. Until I fully do this: https://github.com/cockroachdb/cockroach/issues/131664 it'll be possible for it to regress I think.

**Comment by stevendanna** [comment]
bors r=miraradeva

**Pull request merged**